<a class="anchor" id="0"></a>
# **AdaBoost Classifier Tutorial in Python**


Hello friends,



In recent years, boosting algorithms have gained massive popularity in kaggle competitions. The winners of these competitions use boosting algorithms to achieve high performance. Boosting algorithms such as AdaBoost, Gradient Boosting, and XGBoost are widely used machine learning algorithms. In this kernel, we will discuss **AdaBoost  algorithm**.


So, let's get started.

**If this helped in your learning, then please <font color="red"><b>UPVOTE</b></font>  – because they are the source of motivation!**

<a class="anchor" id="0.1"></a>
# **Notebook Contents**

1. [Intro to Ensemble Machine Learning](#1)
    - [1.1. Bagging](#1.1)
    - [1.2. Boosting](#1.2)
    - [1.3. Stacking](#1.3)
1. [How are base-learners classified](#2)
1. [AdaBoost Classifier](#3)
1. [AdaBoost algorithm intuition](#4)
1. [Difference between AdaBoost and Gradient Boosting model](#5)
1. [AdaBoost implementation in Python](#6)
    - [6.1 Import libraries](#6.1)
    - [6.2 Load dataset](#6.2)
    - [6.3 EDA](#6.3)
    - [6.4 Split dataset into training and test set](#6.4)
    - [6.5 Build the AdaBoost model](#6.5)
    - [6.6 Evaluate model](#6.6)
    - [6.7 Further evaluation with SVC base estimator](#6.7)
1. [Advantages and disadvantages of AdaBoost](#7)
1. [Results and Conclusion](#8)

# **1. Intro to Ensemble Machine Learning** <a class="anchor" id="1"></a>

[Back to Notebook Contents](#0.1)


- An ensemble model is a composite model which combines a series of low performing or weak classifiers with the aim of creating a strong classifier. 

- Here, individual classifiers vote and final prediction label returned that performs majority voting. 

- Now, these individual classifiers are combined according to some specific criterion to create an ensemble model.

- These ensemble models offer greater accuracy than individual or base classifiers. 

- These models can parallelize by allocating each base learner to different mechanisms. 

- So, we can say that ensemble learning methods are meta-algorithms that combine several machine learning algorithms into a single predictive model to increase performance. 

- Ensemble models are created according to some specific criterion as stated below:-

  - **Bagging** - They can be created to decrease model variance using bagging approach.
  
  - **Boosting** - They can be created to decrease model bias using a boosting approach. 
  
  - **Stacking** - They can be created to improve model predictions using stacking approach.
  
  
- It can be depicted with the help of following diagram.

![Ensemble Machine Learning](https://res.cloudinary.com/dyd911kmh/image/upload/f_auto,q_auto:best/v1542651255/image_1_joyt3x.png)

### **1.1 Bagging** <a class="anchor" id="1.1"></a>

- **Bagging** stands for **bootstrap aggregation**. 

- It combines multiple learners in a way to reduce the variance of estimates. 

- For example, random forest trains N Decision Trees where we will train N different trees on different random subsets of the data and perform voting for final prediction. 

- **Bagging ensembles** methods are **Random Forest** and **Extra Trees**.

### ** 1.2 Boosting** <a class="anchor" id="1.2"></a>

- **Boosting** algorithms are a set of the weak classifiers to create a strong classifier. 

- Strong classifiers offer error rate close to 0. 

- Boosting algorithm can track the model who failed the accurate prediction. 

- Boosting algorithms are less affected by the overfitting problem. 

- The following three algorithms have gained massive popularity in data science competitions.

  - AdaBoost (Adaptive Boosting)
  - Gradient Tree Boosting (GBM)
  - XGBoost
  
- We will discuss AdaBoost in this kernel and GBM and XGBoost in future kernels.

- Please refer to my previous kernel - [Bagging vs Boosting](https://www.kaggle.com/prashant111/bagging-vs-boosting?scriptVersionId=24194759)  for a more detailed discussion on on **Bagging** and **Boosting**.

### **1.3 Stacking** <a class="anchor" id="1.3"></a>

- **Stacking** (or stacked generalization) is an ensemble learning technique that combines multiple base classification models predictions into a new data set. 

- This new data are treated as the input data for another classifier. 

- This classifier employed to solve this problem. Stacking is often referred to as blending.

# **2. How are base-learners classified** <a class="anchor" id="2"></a>

[Back to Notebook Contents](#0.1)


- Base-learners are classified into two types.


- On the basis of the arrangement of base learners, ensemble methods can be divided into two groups. 

  - In parallel ensemble methods, base learners are generated in parallel for example - Random Forest. 
  
  - In sequential ensemble methods, base learners are generated sequentially for example AdaBoost.
  

- On the basis of the type of base learners, ensemble methods can be divided into two groups.

  - homogenous ensemble method uses the same type of base learner in each iteration. 
  
  - heterogeneous ensemble method uses the different type of base learner in each iteration.

# **3. AdaBoost Classifier** <a class="anchor" id="3"></a>

[Back to Notebook Contents](#0.1)


- **AdaBoost or Adaptive Boosting** is one of the ensemble boosting classifier proposed by Yoav Freund and Robert Schapire in 1996. 

- It combines multiple weak classifiers to increase the accuracy of classifiers. 

- AdaBoost is an iterative ensemble method. AdaBoost classifier builds a strong classifier by combining multiple poorly performing classifiers so that you will get high accuracy strong classifier. 

- The basic concept behind Adaboost is to set the weights of classifiers and training the data sample in each iteration such that it ensures the accurate predictions of unusual observations. 

- Any machine learning algorithm can be used as base classifier if it accepts weights on the training set. 

- **AdaBoost** should meet two conditions:

   1. The classifier should be trained interactively on various weighed training examples.
  
   2. In each iteration, it tries to provide an excellent fit for these examples by minimizing training error.

- To build a AdaBoost classifier, imagine that as a first base classifier we train a Decision Tree algorithm to make predictions on our training data. 

- Now, following the methodology of AdaBoost, the weight of the misclassified training instances is increased. 

- The second classifier is trained and acknowledges the updated weights and it repeats the procedure over and over again.

- At the end of every model prediction we end up boosting the weights of the misclassified instances so that the next model does a better job on them, and so on.

- AdaBoost adds predictors to the ensemble gradually making it better. The great disadvantage of this algorithm is that the model cannot be parallelized since each predictor can only be trained after the previous one has been trained and evaluated.

- Below are the steps for performing the AdaBoost algorithm:

  1. Initially, all observations are given equal weights.
  
  2. A model is built on a subset of data.
  
  3. Using this model, predictions are made on the whole dataset.
  
  4. Errors are calculated by comparing the predictions and actual values.
  
  5. While creating the next model, higher weights are given to the data points which were predicted incorrectly.
  
  6. Weights can be determined using the error value. For instance,the higher the error the more is the weight assigned to the observation.
  
  7. This process is repeated until the error function does not change, or the maximum limit of the number of estimators is reached.

# **4. AdaBoost algorithm intuition** <a class="anchor" id="4"></a>

[Back to Notebook Contents](#0.1)


- It works in the following steps:

   1. Initially, Adaboost selects a training subset randomly.
  
   2. It iteratively trains the AdaBoost machine learning model by selecting the training set based on the accurate prediction of the last training.
  
   3. It assigns the higher weight to wrong classified observations so that in the next iteration these observations will get the high probability for classification.
  
   4. Also, It assigns the weight to the trained classifier in each iteration according to the accuracy of the classifier. The more accurate classifier will get high weight.
  
   5. This process iterate until the complete training data fits without any error or until reached to the specified maximum number of estimators.
  
   6. To classify, perform a "vote" across all of the learning algorithms you built.
  
  
- The intuition can be depicted with the following diagram:

![AdaBoost Classifier](https://res.cloudinary.com/dyd911kmh/image/upload/f_auto,q_auto:best/v1542651255/image_3_nwa5zf.png)

# **5. Difference between AdaBoost and Gradient Boosting** <a class="anchor" id="5"></a>


[Back to Notebook Contents](#0.1)


- **AdaBoost** stands for **Adaptive Boosting**. It works on sequential ensemble machine learning technique. The general idea of boosting algorithms is to try predictors sequentially, where each subsequent model attempts to fix the errors of its predecessor.


- **GBM or Gradient Boosting** also works on sequential model. Gradient boosting calculates the gradient (derivative) of the Loss Function with respect to the prediction (instead of the features). Gradient boosting increases the accuracy by minimizing the Loss Function (error which is difference of actual and predicted value) and having this loss as target for the next iteration.


- Gradient boosting algorithm builds first weak learner and calculates the Loss Function. It then builds a second learner to predict the loss after the first step. The step continues for third learner and then for fourth learner and so on until a certain threshold is reached.


- So, the question arises in mind that how AdaBoost is different than Gradient Boosting algorithm since both of them works on Boosting technique.


- Both AdaBoost and Gradient Boosting build weak learners in a sequential fashion. Originally, AdaBoost was designed in such a way that at every step the sample distribution was adapted to put more weight on misclassified samples and less weight on correctly classified samples. The final prediction is a weighted average of all the weak learners, where more weight is placed on stronger learners.


- Later, it was discovered that AdaBoost can also be expressed as in terms of the more general framework of additive models with a particular loss function (the exponential loss).


- So, the main differences between AdaBoost and GBM are as follows:-


  1. The main difference therefore is that Gradient Boosting is a generic algorithm to find approximate solutions to the additive modeling problem, while AdaBoost can be seen as a special case with a particular loss function (Exponential loss function). Hence, gradient boosting is much more flexible.


  2. AdaBoost can be interepted from a much more intuitive perspective and can be implemented without the reference to gradients by reweighting the training samples based on classifications from previous learners.


  3. In Adaboost, shortcomings are identified by high-weight data points while in Gradient Boosting, shortcomings of existing weak learners are identified by gradients.


  4. Adaboost is more about ‘voting weights’ and Gradient boosting is more about ‘adding gradient optimization’. 


  5. Adaboost increases the accuracy by giving more weightage to the target which is misclassified by the model. At each iteration, Adaptive boosting algorithm changes the sample distribution by modifying the weights attached to each of the instances. It increases the weights of the wrongly predicted instances and decreases the ones of the correctly predicted instances.

# **6. AdaBoost implementation in Python** <a class="anchor" id="6"></a>

[Back to Notebook Contents](#0.1)


- Now, we come to the implementation part of AdaBoost algorithm in Python.

- The first step is to load the required libraries.

### 6.1 Import libraries <a class="anchor" id="6.1"></a>

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# Any results you write to the current directory are saved as output.

/kaggle/input/iris/Iris.csv
/kaggle/input/iris/database.sqlite


### 6.2 Load dataset <a class="anchor" id="6.2"></a>

In [2]:
iris = pd.read_csv('/kaggle/input/iris/Iris.csv')


### 6.3 EDA <a class="anchor" id="6.3"></a>

### Preview dataset

In [3]:
iris.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

### View summary of dataframe

In [4]:
iris.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
Id               150 non-null int64
SepalLengthCm    150 non-null float64
SepalWidthCm     150 non-null float64
PetalLengthCm    150 non-null float64
PetalWidthCm     150 non-null float64
Species          150 non-null object
dtypes: float64(4), int64(1), object(1)
memory usage: 7.2+ KB


We can see that there are no missing values in the dataset.

### Declare feature vector and target variable

In [5]:
X = iris[['SepalLengthCm','SepalWidthCm','PetalLengthCm','PetalWidthCm']]

X.head()

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
0            5.1           3.5            1.4           0.2
1            4.9           3.0            1.4           0.2
2            4.7           3.2            1.3           0.2
3            4.6           3.1            1.5           0.2
4            5.0           3.6            1.4           0.2

In [6]:
y = iris['Species']

y.head()

0    Iris-setosa
1    Iris-setosa
2    Iris-setosa
3    Iris-setosa
4    Iris-setosa
Name: Species, dtype: object

In [7]:
from sklearn.preprocessing import LabelEncoder

le=LabelEncoder()

y=le.fit_transform(y)

### 6.4 Split dataset into training set and test set <a class="anchor" id="6.4"></a>

In [8]:
# Import train_test_split function
from sklearn.model_selection import train_test_split

# Split dataset into training set and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

### 6.5 Build the AdaBoost model <a class="anchor" id="6.5"></a>

In [9]:
# Import the AdaBoost classifier
from sklearn.ensemble import AdaBoostClassifier


# Create adaboost classifer object
abc = AdaBoostClassifier(n_estimators=50, learning_rate=1, random_state=0)

# Train Adaboost Classifer
model1 = abc.fit(X_train, y_train)


#Predict the response for test dataset
y_pred = model1.predict(X_test)

### Create Adaboost Classifier

- The most important parameters are `base_estimator`, `n_estimators` and `learning_rate`.

- **base_estimator** is the learning algorithm to use to train the weak models. This will almost always not needed to be changed because by far the most common learner to use with AdaBoost is a decision tree – this parameter’s default argument.

- **n_estimators** is the number of models to iteratively train.

- **learning_rate** is the contribution of each model to the weights and defaults to 1. Reducing the learning rate will mean the weights will be increased or decreased to a small degree, forcing the model train slower (but sometimes resulting in better performance scores).

- **loss** is exclusive to AdaBoostRegressor and sets the loss function to use when updating weights. This defaults to a linear loss function however can be changed to square or exponential.

### 6.6 Evaluate Model <a class="anchor" id="6.6"></a>

Let's estimate, how accurately the classifier or model can predict the type of cultivars.

In [10]:
#import scikit-learn metrics module for accuracy calculation
from sklearn.metrics import accuracy_score


# calculate and print model accuracy
print("AdaBoost Classifier Model Accuracy:", accuracy_score(y_test, y_pred))

AdaBoost Classifier Model Accuracy: 0.8666666666666667


- In this case, we got an accuracy of 86.67%, which will be considered as a good accuracy.

### 6.7 Further evaluation with SVC base estimator <a class="anchor" id="6.7"></a>


- For further evaluation, we will use SVC as a base estimator as follows:

In [11]:
# load required classifer
from sklearn.ensemble import AdaBoostClassifier


# import Support Vector Classifier
from sklearn.svm import SVC


# import scikit-learn metrics module for accuracy calculation
from sklearn.metrics import accuracy_score
svc=SVC(probability=True, kernel='linear')


# create adaboost classifer object
abc =AdaBoostClassifier(n_estimators=50, base_estimator=svc,learning_rate=1, random_state=0)


# train adaboost classifer
model2 = abc.fit(X_train, y_train)


# predict the response for test dataset
y_pred = model2.predict(X_test)


# calculate and print model accuracy
print("Model Accuracy with SVC Base Estimator:",accuracy_score(y_test, y_pred))


Model Accuracy with SVC Base Estimator: 0.9111111111111111


- In this case, we have got a classification rate of 91.11%, which is considered as a very good accuracy.


- In this case, SVC Base Estimator is getting better accuracy then Decision tree Base Estimator.

# **7. Advantages and disadvantages of AdaBoost** <a class="anchor" id="7"></a>

[Back to Notebook Contents](#0.1)


- The advantages are as follows:

   1. AdaBoost is easy to implement. 
  
   2. It iteratively corrects the mistakes of the weak classifier and improves accuracy by combining weak learners. 
  
   3. We can use many base classifiers with AdaBoost. 
  
   4. AdaBoost is not prone to overfitting. 

- The disadvantages are as follows:

   1. AdaBoost is sensitive to noise data. 
  
   2. It is highly affected by outliers because it tries to fit each point perfectly. 
  
   3. AdaBoost is slower compared to XGBoost.

# **8. Results and Conclusion** <a class="anchor" id="8"></a>

[Back to Notebook Contents](#0.1)


- In this kernel, we have discussed AdaBoost classifier.

- We have discussed how the base-learners are classified.

- Then, we move on to discuss the intuition behind AdaBoost classifier.

- We have also discuss the differences between AdaBoost classifier and GBM.

- Then, we present the implementation of AdaBoost classifier using iris dataset.

- Lastly, we have discussed the advantages and disadvantages of AdaBoost classifier.

So, now we will come to the end of this kernel.

I hope you find this kernel useful and enjoyable.

Your comments and feedback are most welcome.

Thank you


[Go to Top](#0)